### PR TITLE
Reduce the number of imageDigestExporter step to one per Task

### LINKED ISSUE #1104: Why do we run imageDigestExporter after each step in a taskrun?
The imageDigestExporter container writes the digest of any built images to the container's `terminationMessagePath` which is then parsed by the taskrun controller in order to update the taskrun's status with the image digests.

The container only gets added if a task contains an output resource of type image. However, it [gets added after each step](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/v1alpha1/taskrun/resources/image_exporter.go#L78) as opposed to being added just once (after all the regular steps). The container itself takes as input all the output images that will be produced in a task, and writes the digest for [all of them](https://github.com/tektoncd/pipeline/blob/master/cmd/imagedigestexporter/main.go#L50). So, it is unclear to me why we need to run it after each step. 

Does anyone know why we need to  run the container after each step? Maybe @nader-ziada  or @ImJasonH ?

/kind question

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

Currently, if a `Task` uses one or more image output we add a
`imageDigestExporter` for each step defined in the `Task`. This is not
required as only one step/container can do the work and handle
multiple images.

This change reduce to one the number of `imageDigestExporter` to one
step, after the last user-defined step.

Closes #1104 

Signed-off-by: Vincent Demeester <vdemeest@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) and [TaskRun](../tekton/publish-run.yaml) to build and release this image

# Release Notes

```
Reduce the number of imageDigestExporter step to one per Task instead of one for each steps in a Task
```

### CODE REVIEW

**Comment by vdemeester** [comment]
/hold
I feel we need a e2e test with multiple images :angel:

**Comment by bobcatfish** [comment]
I'm trying to remember the details of the reason we had an image after each step - I even paired with @nader-ziada on this and I can't quite remember 🤣 I'm really kicking myself that we didn't make it clear in comments :S

Part of the reason was that it seemed like we couldn't guarantee that we knew which step was exporting the image - maybe to handle a case where we export 2 images, each in a different step??  But it sounds like we're handling that already - all I can come up with is that maybe it was to handle a case where the same image is getting exported multiple times but I'm not sure now why we'd want to support that 🤷‍♀ 

Might as well go ahead with this for now unless @nader-ziada is around and can remember why we did it? 😅

**Comment by nader-ziada** [comment]
Agree with @bobcatfish it had to do with building same image in multiple steps, but makes sense to me to have just one step in the end.

**Comment by vdemeester** [comment]
> This looks good to me, but I feel we need to document what type of restrictions we have in terms of using the image pipeline resources and build the output.
> I think it makes sense to have a Task producing one image - multi-stage docker files can be used where needed, so in general only one output of type image should be available in a Task.
> It's fine to do that as a follow-up.

Well, it's possible for a task to build and push multiple images :angel:

File: pkg/reconciler/v1alpha1/taskrun/resources/image_exporter.go [hunk 1]
```diff
@@ -75,10 +75,8 @@ func AddOutputImageDigestExporter(
 				return xerrors.Errorf("Failed to format image resource data for output image exporter: %w", err)
 			}
 
-			for _, s := range taskSpec.Steps {
-				augmentedSteps = append(augmentedSteps, s)
-				augmentedSteps = append(augmentedSteps, imageDigestExporterContainer(s.Name, imagesJSON))
-			}
+			augmentedSteps = append(augmentedSteps, taskSpec.Steps...)
+			augmentedSteps = append(augmentedSteps, imageDigestExporterContainer(imagesJSON))
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/resources/image_exporter.go]: at this point we probably don't even need `augmentedSteps` ? could use `taskSpec.Steps` directly

**Pull request merged**